src/main/java/build/buildfarm/worker/CASFileCache.java

```diff
@@ -272,21 +272,23 @@ private void fetchDirectory(
     }
   }
 
-  public synchronized Path putDirectory(
+  public Path putDirectory(
       Digest digest,
       Map<Digest, Directory> directoriesIndex) throws IOException, InterruptedException {
     Path path = getDirectoryPath(digest);
 
-    DirectoryEntry e = directoryStorage.get(digest);
-    if (e != null) {
-      incrementReferences(e.inputs);
-      return path;
+    synchronized (this) {
+      DirectoryEntry e = directoryStorage.get(digest);
+      if (e != null) {
+        incrementReferences(e.inputs);
+        return path;
+      }
     }
 
     ImmutableList.Builder<Path> inputsBuilder = new ImmutableList.Builder<>();
     fetchDirectory(digest, path, digest, directoriesIndex, inputsBuilder);
```

Comment:
So the comment on the fetchDirectory function is not true? Why, what changed? It does look like it needs sync to me.

Reply:
The comment is now extraneous - I don't see any issues with calling fetchDirectory from an cache-wide-unsynchronized context. There is, however a race when two misses on the same key occur, and when similar things happen to files via put - I will have to add a path-based synchronization from a separate unit of work.

Reply:
Will do. The synchronization is largely academic here, since we only do puts of files or directories from the single threaded input stage in the operationqueue worker. That's about to shift substantially...